// eslint-config-dustinspecker/index.js
'use strict';
module.exports = {
  env: {
    node: true
  },
  rules: {
    'array-bracket-spacing': [2, 'never'],
    'block-scoped-var': 2,
    'brace-style': [2, '1tbs', {allowSingleLine: false}],
    camelcase: 2,
    'comma-dangle': [2, 'never'],
    'comma-spacing': [2, {
      after: true,
      before: false
    }],
    'comma-style': 0,
    complexity: [2, 17],
    'computed-property-spacing': [2, 'never'],
    'consistent-return': 2,
    'consistent-this': [2, 'self'],
    curly: 2,
    'default-case': 2,
    'dot-notation': 2,
    eqeqeq: 2,
    'eol-last': 2,
    'func-names': 0,
    'func-style': [2, 'declaration'],
    'guard-for-in': 2,
    'handle-callback-err': 2,
    indent: [2, 2, {SwitchCase: 1, VariableDeclarator: 2}],
    'key-spacing': [2, {
      afterColon: true,
      beforeColon: false
    }],
    'max-nested-callbacks': [2, 5],
    'new-cap': [2, {
      capIsNew: true,
      newIsCap: true
    }],
    'new-parens': 2,
    'no-alert': 2,
    'no-array-constructor': 2,
    'no-call
