// tap-mocha-reporter/lib/formatter.js
// A formatter is a Duplex stream that TAP data is written into,
// and then something else (presumably not-TAP) is read from.
//
// See tap-classic.js for an example of a formatter in use.

var Duplex = require('stream').Duplex
if (!Duplex) {
  try {
    Duplex = require('readable-stream').Duplex
  } catch (er) {
    throw new Error('Please install "readable-stream" to use this module ' +
                    'with Node.js v0.8 and before')
  }
}

var util = require('util')
var Parser = require('tap-parser')
util.inherits(Formatter, Duplex)
module.exports = Formatter

function Formatter(options, parser, parent) {
  if (!(this instanceof Formatter))
    return new Formatter(options, parser, parent)

  if (!parser)
    parser = new Parser()

  Duplex.call(this, options)
  this.child = null
  this.parent = parent || null
  this.level = parser.level
  this.parser = parser

  attachEvents(this, parser, options)

  if (options.init)
    options.init.call(this)
}

function attachEvents (self, parser, options) {
  var events = [
    'version', 'plan', 'assert', 'comment',
    'complete', 'extra', 'bailout'
  ]

  parser.on('child', function (childparser) {
    self.child = new Formatter(options, childparser, self)
    if (options.child)
      options.child.call(self, self.child)
  })

  events.forEach(function (ev) {
    if (typeof options[ev] === 'function')
      parser.on(ev, options[ev].bind(self))
  })

  // proxy all stream events directly
  var streamEvents = [
    'pipe', 'prefinish', 'finish', 'unpipe', 'close
